probe-rs/src/architecture/riscv/mod.rs

```diff
@@ -194,22 +195,151 @@ impl<'state> Riscv32<'state> {
 
         Ok(tselect_index)
     }
+
+    /// Reads the trigger data from the target and stores it in the state.
+    fn load_trigger_data(&mut self) -> Result<Vec<Trigger>, Error> {
+        // this can be called w/o halting the core via Session::new - temporarily halt if not halted
+        let was_running = !self.core_halted()?;
+        if was_running {
+            self.halt(Duration::from_millis(100))?;
+        }
+        let bp_count = self.determine_number_of_hardware_breakpoints()?;
+
+        let tselect = 0x7a0;
+        let tdata1 = 0x7a1;
+        let tdata2 = 0x7a2;
+
+        // Load the initial data from the trigger units
+        let mut breakpoints = vec![];
+        let num_hw_breakpoints = bp_count as usize;
+        for bp_unit_index in 0..num_hw_breakpoints {
+            // Select the trigger.
+            self.write_csr(tselect, bp_unit_index as u32)?;
+
+            // Read the trigger "configuration" data.
+            let tdata_value = Mcontrol(self.read_csr(tdata1)?);
+
+            tracing::debug!("Breakpoint {}: {:?}", bp_unit_index, tdata_value);
+
+            // The trigger must be active in at least a single mode
+            let trigger_any_mode_active = tdata_value.m() || tdata_value.s() || tdata_value.u();
+
+            // Only return if the trigger if it is for an execution debug action in all modes.
+            if tdata_value.type_() == 2
+                && tdata_value.action() == 1
+                && tdata_value.match_() == 0
+                && trigger_any_mode_active
+                && tdata_value.execute()
+            {
+                let breakpoint = self.read_csr(tdata2)?;
+                breakpoints.push(Trigger::AddressMatch(breakpoint));
+            } else {
+                breakpoints.push(Trigger::None);
+            }
+        }
+
+        if was_running {
+            self.resume_core()?;
+        }
+
+        Ok(breakpoints)
+    }
+
+    /// Returns a mutable slice of the state of all trigger units.
+    ///
+    /// This function loads the trigger data from the target if it has not been loaded yet.
+    fn triggers_mut(&mut self) -> Result<&mut [Trigger], Error> {
+        if self.state.hw_breakpoints.is_none() {
+            let breakpoints = self.load_trigger_data()?;
+            self.state.hw_breakpoints = Some(breakpoints);
+        }
+
+        Ok(self.state.hw_breakpoints.as_deref_mut().unwrap())
+    }
+
+    /// Returns the state of all trigger units.
+    fn triggers(&mut self) -> Result<&[Trigger], Error> {
+        Ok(self.triggers_mut()?)
+    }
+
+    fn disable_trigger_on_target(&mut self, tsel: u32) -> Result<(), RiscvError> {
+        let tselect = 0x7a0;
+        let tdata1 = 0x7a1;
+        let tdata2 = 0x7a2;
```

Comment:
May I suggest that these magic numbers are replaced by meaningful names defined in an enum?

Reply:
We can massage the code, sure, but how much more meaningful a name would you like than `tselect` which is the name of the register whose address is `0x7a0`?

Reply:
Architecturally, the CSR address represented by tselect is never going to change.  As currently written, the code offers 3 opportunities to get it wrong. Using an enum name means you only have to get it right once.